src/ethereum/cancun/vm/__init__.py

```diff
@@ -58,6 +54,8 @@ class Message:
     """
 
     caller: Address
+    origin: Address
```

Comment:
The rule I've been loosely following (not that I've modified either class recently 😅) is that things remaining constant during a transaction belong in `Environment` and things that can change per-call belong in `Message`.

What is your proposed new rule for placing new members?

Reply:
The rules I have followed are

- Block scoped items go into `Environment`. These are constants over a block. Examples would be anything chain specific or block header specific (chain_id, timestamp etc.). This way, we don't have to re-compute `env` for each transaction.
- Transaction scoped items go into `Message`. These could be constant over calls (e.g. origin, gas_price etc.) or change (e.g. depth, target etc.)

However, I do see value in separating constants over a transaction from those things that change per call.
So may be good idea to have the following
- `BlockEnvironment` - For block scope items
- `TransactionEnvironment` - For constants over a tx
- `Message` - For items that change per message call. This can have an instance of `block_env` and `tx_env`

I feel like this will give us an even cleaner and more logical separation of concerns. What do you think of that idea?

Reply:
PS: The dataclass names are just a suggestion :). Totally happy to have some other names

Reply:
> The rules I have followed are [...]

That makes a lot of sense! I'm happy with that.

> However, I do see value in separating constants over a transaction from those things that change per call.

This sounds like a big refactor. Do you think it's worth it?

Reply:
I do believe its worth it. I just pushed a few more commits on the Cancun fork to demonstrate the benefits.

1. This allows us to neatly separate out the logic for processing an entire block vs processing individual transactions. So if we want to process a new block, we just run the `apply_body` function and if we want to process an individual transaction, we just run the `process_transaction` function.
2. Point 1 above makes it really clear which components need to be updated when someone is implementing a new EIP. Right now, I feel like its still a bit vague.
3. This change greatly simplifies our `t8n` tool, removing the need for a majority of the fork specific conditionals. See [here](https://github.com/gurukamath/execution-specs/blob/10f3a2766a076afb51580709165cba4fca4efd7f/src/ethereum_spec_tools/evm_tools/t8n/__init__.py#L325). So for state tests we essentially just call `process_transaction` and for blockchain tests, we just call the `apply_body` function. And the signatures of these functions are unlikely to change greatly from fork to fork.
4. Simplifies some of our tracing logic by providing us a vehicle to store traces over an entire transaction. This also makes it easier to move to writing traces immediately to file (as opposed to status quo where we store them all and output them in the end).